drafts/sorting-data-with-clojure.md

```diff
@@ -0,0 +1,140 @@
+
+That gets you a new vector, with the smallest invoices first and the valuable ones at the end. Hardly useful for business, so lets flip it around by supplying a comparator function too:
+
+```clojure
+(sort-by > :invoice/total-before-tax invoices)
+```
+
+Now we’re cooking with gas! The most valuable invoices are now at the head of the list! Need the top 10? Just `(take 10)` and you’re set.
+
+How did this happen? Clojure compared the values returned by `:invoice/total-before-tax` using the `>` function.
+
+## Sorting by composite keys
+
+The next requirement might be to sort by `:invoice/number` and `:invoice/total-before-tax`. Imagine the idea is that when two invoices have the same total that they are then sorted by their invoice number to show some kind of implied order.
+
+This is where our friend `(juxt)` comes in. `(juxt)` accepts a list of functions and returns a new function, that when called, returns the results of all the original functions in a vector.
+
+```clojure
+(def head-and-tail (juxt [first last]))
+(head-and-tail [1 2 3 4 5]) #=> [1 5]
+```
+
+Here you can see that juxt applied `(first)`, and `(last)`, to the supplied list of numbers and gave us the head and the tail of the list. Keen readers might have just figured out where I’m going with this.
+
+`(sort-by)` can compare these vectors too, so we can sort our invoices like this:
+
+```clojure
+(def total-before-tax-and-number (juxt :invoice/total-before-tax :invoice/number))
+(sort-by total-before-tax-and-number invoices)
+```
+
+Now the results will be sorted from lowest value invoice to the highest, with the invoice numbers in order too. So we're halfway there.
+
+## Mixing the order (or composing functions)
+
+The results of the previous example doesn't make much sense. How can we combine sorting the invoice amounts in descending order _and_ have the invoice numbers run sequentially when there is an overlap?
+
+One possible solution is to use `(comp)`, and make a new function that will return the value of `:invoice/total-before-tax` as a negative number. `(comp)` works by accepting a list of functions and returning a new function, which when called, calls the arguments from right to left and passing the result of the previous call to the next one, starting with the parameter when called.
+
+An example will be worth a thousand words:
+
+```clojure
+(require '[clojure.string :as str])
+(def up-and-reverse (comp str/reverse str/upper-case))
+(up-and-reverse "elloh") #=> "HELLO"
+
+;; or
+
+(str/reverse (str/upper-case "elloh")) => "HELLO"
+```
+
+In order to get the negative total we can just `(comp)` together `(-)` and `:invoice/total-before-tax` like this:
+
+```clojure
+(def negative-total (comp - :invoice/total-before-tax))
+(negative-total invoice) #=> -100
+```
+
```

Comment:
If this right-to-leftness of `comp` bothers you, you could also simply declare it as an anonymous function which wraps a thread-first functional pipeline: `#(-> % :invoice/total-before-tax -)`.